### PR TITLE
FLINK-3405: Extend NiFiSource with interface StoppableFunction

### PR DESCRIPTION
**Pull request by smarthi**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-3405] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by smarthi** [comment]
I started to add some NiFi integration tests which are presently lacking, but realized that it would be better to have that as part of a different jira.

File: flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java [hunk 1]
```diff
@@ -143,4 +144,13 @@ public void close() throws Exception {
...
 	}
 
+ /**
+	* {@inheritDoc}
+	* <p>
+	* Sets the {@link #isRunning} flag to {@code false}.
+	*/
+	@Override
+	public void stop() {
+		this.isRunning = false;
```
Comment (zentol) [inline flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java]: I'd like other to chime in here: do you have to call close() within stop()? The Twitter source is doing this.

Comment (smarthi) [inline flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java]: I guess we need to for now, until stop() is wired into the StreamGraph i.e. Thanks for pointing that out.

Comment (StephanEwen) [inline flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java]: Stop should not call close(), actually. It should just cause the main source loop to gracefully exit.
Close() will happen anyways after that.

Comment (zentol) [inline flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java]: @StephanEwen could you point me to the code where close() is called for a stopped source?

Comment (zentol) [inline flink-streaming-connectors/flink-connector-nifi/src/main/java/org/apache/flink/streaming/connectors/nifi/NiFiSource.java]: nvm, found it.

**Comment by StephanEwen** [comment]
close() is called by all operators when the operator finishes without an exception of canceling. Have a look at the AbstractUDFStream

Stop() is thought to be merely a trigger to make an unbounded source stop its unbounded loop. After that, it should finish like a bounded source, including the close() call.

**Comment by StephanEwen** [comment]
I would do one change here: Make the `isRunning` flag non transient, and set it to true initially, rather than in open.

That can act as a safeguard against very early stopping (before open is even completed), were stop would set the flag to false, and open would set it back to true.

I think the runtime does prevent open to happen after stop, but cannot hurt to guard this in the source as well.

**Comment by StephanEwen** [comment]
All right, merging this, thanks!

**Pull request closed without merging**